swift/ql/src/queries/Security/CWE-327/BrokenCryptoAlgorithm.ql

```diff
@@ -0,0 +1,30 @@
+/**
+ * @name Use of a broken or weak cryptographic algorithm
+ * @description Using broken or weak cryptographic algorithms can compromise security.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 7.5
+ * @precision high
+ * @id swift/weak-cryptographic-algorithm
+ * @tags security
+ *       external/cwe/cwe-327
+ */
+
+import swift
+import codeql.swift.security.Cryptography // TODO: or use a megamodule like Concepts, containing a Cryptography module?
+
+from
+  Cryptography::CryptographicOperation operation, Cryptography::CryptographicAlgorithm algorithm,
+  string msgPrefix
+where
+  algorithm = operation.getAlgorithm() and
+  // `Cryptography::HashingAlgorithm` and `Cryptography::PasswordHashingAlgorithm` are
+  // handled by `py/weak-sensitive-data-hashing`
+  algorithm instanceof Cryptography::EncryptionAlgorithm and
+  (
+    algorithm.isWeak() and
+    msgPrefix = "The cryptographic algorithm " + operation.getAlgorithm().getName()
+  )
+  or
+  operation.getBlockMode().isWeak() and msgPrefix = "The block mode " + operation.getBlockMode()
```

Comment:
There may be some overlap in results here with the `swift/ecb-encryption` query - though that query currently only models a very narrow range of algorithms and does not use heuristics.  We might want to sort this out later.  But I suggest we make it a follow-up issue and don't let it distract from finishing the `swift/weak-cryptographic-algorithm` query.